src/corelibs/U2View/src/ov_msa/exclude_list/MsaExcludeList.cpp

```diff
@@ -0,0 +1,593 @@
+        sequences << getExcludeListRowSequence(listItem);
+        excludeListRowIdsMovedToMsa << getExcludeListRowId(listItem);
+    }
+    removeEntries(selectedItems);
+
+    int versionBefore = msaObject->getObjectVersion();
+    int insertionIndex = -1;  // Append by default.
+    if (!msaEditor->getSelection().isEmpty() && msaEditor->getRowOrderMode() == MaEditorRowOrderMode::Original) {
+        insertionIndex = msaEditor->getSelection().getRectList().last().bottom() + 1;
+    }
+    AddSequenceObjectsToAlignmentTask task(msaObject, sequences, insertionIndex, true);
+    task.run();  // TODO: extract as an utility method.
+    if (!task.hasError()) {
+        trackedUndoMsaVersions.insert(versionBefore, {false, excludeListRowIdsMovedToMsa});
+        trackedRedoMsaVersions.insert(msaObject->getObjectVersion(), {false, excludeListRowIdsMovedToMsa});
+    }
+    updateState();
+}
+
+void MsaExcludeListWidget::handleUndoRedoInMsaEditor(const MultipleAlignment &maBefore, const MaModificationInfo &modInfo) {
+    auto msaObject = msaEditor->getMaObject();
+    int version = msaObject->getObjectVersion();
+    if (modInfo.type != MaModificationType_Undo && modInfo.type != MaModificationType_Redo) {
+        auto truncateVersionData = [version](QMap<int, UndoRedoStep> &undoRedoMap) {
+            CHECK(!undoRedoMap.isEmpty() && version <= undoRedoMap.lastKey(), );
+            QList<int> allUndoRedoVersions = undoRedoMap.keys();
+            for (int undoRedoVersion : qAsConst(allUndoRedoVersions)) {
+                if (undoRedoVersion >= version) {
+                    undoRedoMap.remove(undoRedoVersion);
+                }
+            }
+        };
+        truncateVersionData(trackedUndoMsaVersions);
+        truncateVersionData(trackedRedoMsaVersions);
+        return;
+    }
+    bool isRedo = modInfo.type == MaModificationType_Redo;
+    CHECK((isRedo && trackedRedoMsaVersions.contains(version)) || (!isRedo && trackedUndoMsaVersions.contains(version)), );
+
+    const UndoRedoStep &undoRedoContext = isRedo ? trackedRedoMsaVersions.value(version) : trackedUndoMsaVersions.value(version);
+    bool isAddToExcludeList = (isRedo && undoRedoContext.isMoveFromMsaToExcludeList) || (!isRedo && !undoRedoContext.isMoveFromMsaToExcludeList);
+    if (isAddToExcludeList) {  // Add rows removed from MSA to Exclude list
+        QList<MultipleAlignmentRow> msaRows;
+        QSet<qint64> msaRowIdsAfter = msaObject->getRowIds().toSet();
+        for (int i = 0; i < maBefore->getNumRows(); i++) {
+            const MultipleAlignmentRow &row = maBefore->getRow(i);
+            if (!msaRowIdsAfter.contains(row->getRowId())) {
+                msaRows << row;
+            }
+        }
+        SAFE_POINT(msaRows.size() == undoRedoContext.excludeListRowIdsDelta.size(), "Failed to map Exclude List rows to Msa rows", )
+        for (int i = 0; i < msaRows.size(); i++) {
+            auto &msaRow = msaRows[i];
```

Comment:
Maybe `const auto&`?

Reply:
It won't change anything: the list is not shared (is local). Added 'const' to move forward.